test/kubernetes_unittest.cc

```diff
@@ -1905,4 +1917,63 @@ TEST_F(KubernetesTestFakeServerOneWatchRetryClusterLevelMetadata,
   server->TerminateAllStreams();
 }
 
+namespace {
+class FakeKubernetesUpdater : public KubernetesUpdater {
+ public:
+  FakeKubernetesUpdater(const Configuration& config,
+                        HealthChecker* health_checker,
+                        MetadataStore* store)
+    : KubernetesUpdater(
+          config, health_checker, store,
+          WaitableTimerFactoryImpl<
+              testing::FakeClock,
+              testing::WaitTraits<testing::FakeClock>>::New()) {}
+};
+}
+
+TEST_F(KubernetesTestFakeServerThreeWatchRetriesNodeLevelMetadata,
+       KubernetesUpdaterHourlyReconnection) {
+  const std::string nodes_watch_path =
+    "/api/v1/watch/nodes/TestNodeName?watch=true";
+  const std::string pods_watch_path =
+    "/api/v1/pods?fieldSelector=spec.nodeName%3DTestNodeName&watch=true";
+
+  // Create a fake server representing the Kubernetes master.
+  server->SetResponse("/api/v1/nodes?limit=1", "{}");
+  server->SetResponse("/api/v1/pods?limit=1", "{}");
+  server->AllowStream(nodes_watch_path);
+  server->AllowStream(pods_watch_path);
+
+  MetadataStore store(*config);
+  FakeKubernetesUpdater updater(*config, /*health_checker=*/nullptr, &store);
+  updater.Start();
+
+  // Wait for connection #1.
+  const auto timeout = time::seconds(3);
+  ASSERT_TRUE(server->WaitForStreamWatchers(nodes_watch_path, 1, timeout));
+  ASSERT_TRUE(server->WaitForStreamWatchers(pods_watch_path, 1, timeout));
+
+  // Advance fake clock only 30 min, not enough to trigger reconnection.
+  testing::FakeClock::Advance(std::chrono::seconds(1800));
+  EXPECT_EQ(1, server->NumWatchers(nodes_watch_path));
+  EXPECT_EQ(1, server->NumWatchers(pods_watch_path));
+
+  // Advance another 30 min to trigger connection #2.
+  //
+  // NOTE: The FakeServer doesn't clean up state when client
```

Comment:
Why not? Should we just add that functionality?

Reply:
I applied the same changes here, so either PR can go in first.